src/neon/neon_radix4.rs

```diff
@@ -0,0 +1,489 @@
+                len,
+                Neon32Butterfly::Len1(NeonF32Butterfly1::new(direction)),
+            ),
+            1 => (
+                len,
+                Neon32Butterfly::Len2(NeonF32Butterfly2::new(direction)),
+            ),
+            2 => (
+                len,
+                Neon32Butterfly::Len4(NeonF32Butterfly4::new(direction)),
+            ),
+            3 => (
+                len,
+                Neon32Butterfly::Len8(NeonF32Butterfly8::new(direction)),
+            ),
+            _ => {
+                if num_bits % 2 == 1 {
+                    if len < USE_BUTTERFLY32_FROM {
+                        (8, Neon32Butterfly::Len8(NeonF32Butterfly8::new(direction)))
+                    } else {
+                        (
+                            32,
+                            Neon32Butterfly::Len32(NeonF32Butterfly32::new(direction)),
+                        )
+                    }
+                } else {
+                    (
+                        16,
+                        Neon32Butterfly::Len16(NeonF32Butterfly16::new(direction)),
+                    )
+                }
+            }
+        };
+
+        // precompute the twiddle factors this algorithm will use.
+        // we're doing the same precomputation of twiddle factors as the mixed radix algorithm where width=4 and height=len/4
+        // but mixed radix only does one step and then calls itself recusrively, and this algorithm does every layer all the way down
+        // so we're going to pack all the "layers" of twiddle factors into a single array, starting with the bottom layer and going up
+        let mut twiddle_stride = len / (base_len * 4);
+        let mut twiddle_factors = Vec::with_capacity(len * 2);
+        while twiddle_stride > 0 {
+            let num_rows = len / (twiddle_stride * 4);
+            for i in 0..num_rows / 2 {
+                for k in 1..4 {
+                    unsafe {
+                        let twiddle_a = twiddles::compute_twiddle::<f32>(
+                            2 * i * k * twiddle_stride,
+                            len,
+                            direction,
+                        );
+                        let twiddle_b = twiddles::compute_twiddle::<f32>(
+                            (2 * i + 1) * k * twiddle_stride,
+                            len,
+                            direction,
+                        );
+                        let twiddles_packed =
+                            vld1q_f32(&[twiddle_a, twiddle_b] as *const _ as *const f32);
+                        twiddle_factors.push(twiddles_packed);
+                    }
+                }
+            }
+            twiddle_stride >>= 2;
+        }
+
+        // make a lookup table for the bit reverse shuffling
+        let rest_len = len / base_len;
+        let bitpairs = (rest_len.trailing_zeros() / 2) as usize;
+        let shuffle_map = (0..rest_len)
+            .map(|val| reverse_bits(val, bitpairs))
```

Comment:
I bet the bit reversal itself could work as a SIMD algorithm. Obviously not something to do for this PR, but could be interesting future research.